### ISSUE TITLE
Cancelled composer cli commands can leave the /vendor/ folder in a inconsistent state

### ISSUE THREAD

**Issue by staabm** [composer/composer#4736]
I invested several hours to find the cause of a "class not found error".

The cause for this error was a user cancelled `composer update` run using`ctrl+C`.
we didnt realize that some folder were left back empty/inconsistent and so autoloading didnt work because of `file_exists` of the `Classloader` was not able to succeed.

I guess we have 2 options to enhance DX in such a case:
1) best: make cli commands (especially `composer update`) automic, e.g. by using a working directory and copy everything in place when download etc. finished.
2) when a CLI command is interrupted via a signal, e.g. `ctrl+c`, `ctrl+d`,... output a warning that this might lead to inconsistent autoloading/behaviour/whatever

The actual "issue" was to find the reason why it doesnt work. `composer validate`, `composer -vvv install`, `composer -vvv dump-autoload` did not show up a error...

**Comment by stof** [comment]
Your first option is not good either: a filesystem copy is not atomic either.

**Comment by staabm** [comment]
> Your first option is not good either: a filesystem copy is not atomic either.

how to get the operation atomic is the problem to solve right. If we use a symlink or rename or something like that needs to be discussed, right.

**Comment by staabm** [comment]
An alternative which came into my mind: before starting work/commands which will lead to inconsistencies if interrupted, we could place marker somewhere which indicates that there is some work in progress. This marker would be deleted in case the work was completed.

Some (All?) commands could check for this indicator before they start work. In case the indicator is present, a warning/error is printed with some guidance on how to resolve the inconsistency.

This of course would just be a workaround in case we dont get things at least "kind of atomic".

**Comment by curry684** [comment]
Atomicity isn't an option. Even if you do all the work in `vendor.temp`, then rename `vendor` to `vendor.old` and then `vendor.temp` to `vendor` before deleting `vendor.old` recursively and writing `composer.lock` and `composer.json` you may end up with, depending on the timing of the ctrl-C, or assuming the worst a power loss):

1. No `vendor` directory at all
2. A leftover `vendor.temp` waiting to be forgotten
3. A partially deleted `vendor.old` waiting to be forgotten
4. An inconsistent `composer.lock`
5. Depending on the operation an inconsistent `composer.json`

All this is at the cost of your web project temporarily taking up double disk space, and let's not forget most hosting servers have strict disk quota in place. You wouldn't want to have the entire Sonata suite of several hundreds of MB hanging around multiple times if you're tight on space.

A lockfile mechanism could help, but I can't think of any good recovery mechanisms. The only proper solution to any kind of inconsistency going to be `rm -rf vendor && composer install` to revert to the last known good state. So the only point of the lockfile is going to be printing a message to that effect - you would want to automatically perform such a destructive action.

**Comment by staabm** [comment]
What do you guys think of adding emitting a WARNING with instructions on how to "repair" the maybe incosistent `vendor/` dir on `ctrl+c` interrupted composer processes?

**Comment by Seldaek** [comment]
I think warning on ctrl+c if it is done during Composer\Installer's operation execution phase would probably be a good thing. It won't save you but might give you a hint that you just messed up :) TBH though the easiest way to keep going is to just run the command you ran again. That'll generally fix things as far as I can think of as we only write the new state to the installed.json after an operation is completed. That means if a package is partially downloaded or partially updated it'll still be seen as old and needing an update/install/...

**Comment by curry684** [comment]
Can PHP code even detect and act on ctrl-C signals? I think it's just a pkill from that end, and we might only be able to intervene with a shutdown handler, if at all. Which would need a lockfile of sorts to detect the abnormal situation, at which point we're pretty much 90% through the entire detection mechanism. Still not really sure all of it is worth the effort.

In another issue we did discuss that the semi-atomic in-place replacement method has merits of its own because it reduces downtime of the site. Whether we like it or not people are using `composer install` and even `update` on live websites, and the replace method would not only reduce downtimes to practically zero, but also not result in a broken site if it fails for example because of borked dependencies. So imo that path is worth debate, and it would solve all of this along the way.

**Comment by Seldaek** [comment]
Yeah we can detect that https://github.com/composer/composer/blob/master/src/Composer/Command/CreateProjectCommand.php#L300

Atomic ops sounds quite horrible on dev boxes for performance and wtf reasons (e.g. you have open files in the vendor and then the whole dir gets swapped, some editors might not like it). It would be a small benefit for people that do it live in prod sure, but those people are doing deployments very wrong IMO. It's probably fine for them to have a couple seconds of downtime.. I do it myself on non-business-critical sites because tbh in some cases it's not worth the hassle. IMO this is not worth the trouble.

**Comment by curry684** [comment]
That detection is hacky on *nix and non-existent on Windows. Also, even on *nix it's nowhere [guaranteed to be present](http://php.net/manual/en/pcntl.installation.php):

> Process Control support in PHP is not enabled by default. You have to compile the CGI or CLI version of PHP with --enable-pcntl configuration option when compiling PHP to enable Process Control support.

Just tried on an up to date DirectAdmin server and it doesn't have it:
```
$ php -r "var_dump(function_exists('pcntl_signal'));"
bool(false)
```
Think as a whole it's a 'not worth the effort' then.

**Comment by Seldaek** [comment]
Yeah maybe it is.. overall it's not really a big gain anyway.
